File: 322-coin-change/coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        // initialize array = infty of size 0:amount
        // array[0] = 0
        // for amt in 1:amount
        // array[amt] = min{coinChange(amt-coin) + 1 | coin in coins, amt-coin >= 0, coinChange(amt-coin) != infty}
        int[] array = new int[amount + 1];
        for (int i = 0; i <= amount; i++) {
            array[i] = amount + 1;
        }
        array[0] = 0;
        for (int i = 1; i <= amount; i++) {
            for (int coin : coins) {
                if ((i - coin >= 0) && (array[i-coin] != amount + 1)) {
                    if (array[i-coin] + 1 < array[i]) {
                        array[i] = array[i-coin] + 1;
                    }
                }
            }
        }

        int res = array[amount];
        if (res == amount + 1) {
            return -1;
        }

        return res;
    }
}

File: 322-coin-change/coin-change.py
class Solution(object):
    def coinChange(self, coins, amount):
        """
        :type coins: List[int]
        :type amount: int
        :rtype: int
        """

        # s[i] = smallest needed to make amount i
        # s[i] = min({i - s[i-amt] for amt in coins}) + 1
        amounts = [0]
        for i in range(1, amount + 1):
            valid_coins = [c for c in coins if c <= i]
            candidates = [amounts[i - c] + 1 for c in valid_coins if amounts[i - c] >= 0]
            if len(candidates) == 0:
                amounts.append(-1)
            else:
                amounts.append(min(candidates))
        return amounts[amount]
        